### PR TITLE
WIP Async2

### PR DESCRIPTION
**Pull request by majocha**
Currently just replaces cancellable.
Features: cold start, multi-start, cancellable, trampolined binds for safe non-tail recursion, fast tail-recursion with `ReturnFromFinal`, exceptions cached in ExceptionDispatchInfo for fast recovery from deep call stacks.

Synchronous executions should be quite fast and low alloc when it's done, but there are some fixes needed:
Currently complex mut rec expressions are not statically compiled, see the nowarn in  `CheckDeclarations.fs`.

Very initial work.

### CODE REVIEW

**Comment by vzarytovskii** [comment]
To avoid confusion - it's not the same "async2" as in runtime-supported async. It's just different re-implementation of async, right?

**Comment by majocha** [comment]
> To avoid confusion - it's not the same "async2" as in runtime-supported async. It's just different re-implementation of async, right?

Nope, no. I wasn't even aware this was a thing until now :) The goal is to possibly replace all the internal use of async and cancellable with one reasonably efficient resumable CE. I went with async2 name after the example in resumable code RFC.

LOL, probably before I ever finish this, the runtime async model will switch to https://github.com/dotnet/runtimelab/blob/feature/async2-experiment/docs/design/features/runtime-handled-tasks.md

I guess this is the fate, always catching up :)

**Comment by majocha** [comment]
I updated more places where previously `async` (and `node` before)  where in use. Particularly the `GraphNode` async lazy type.
F# inlining makes it hard to test the internal types. `InternalsVisibleTo` will not work in such case.
I had to actually link the relevant source files from the FCS to the test project. `Async2.fs` and `BuildGraph.fs` are now compiled also inside the test project to make the types accessible in tests.

File: tests/FSharp.Compiler.ComponentTests/FSharp.Compiler.ComponentTests.fsproj [hunk 1]
```diff
@@ -352,6 +346,18 @@
...
     <Compile Include="Attributes\AttributeCtorSetPropAccess.fs" />
+    <Compile Include="CompilerService\RangeModule.fs" />
+    <Compile Include="CompilerService\Caches.fs" />
+    <Compile Include="CompilerService\LruCache.fs" />
+    <Compile Include="..\..\src\Compiler\Utilities\Async2.fs">
+        <Link>CompilerService\Async2.fs</Link>
+    </Compile>
+    <Compile Include="..\..\src\Compiler\Facilities\AsyncMemoize.fs">
+        <Link>CompilerService\AsyncMemoImpl.fs</Link>
+    </Compile>
```
Comment (majocha) [inline tests/FSharp.Compiler.ComponentTests/FSharp.Compiler.ComponentTests.fsproj]: Linked sources like these seem to bring down VS to a crawl.

Comment (majocha) [reply tests/FSharp.Compiler.ComponentTests/FSharp.Compiler.ComponentTests.fsproj]: This is probably related to in memory cross proj references. Project options manager goes into a loop of invalidating projects and type checks everything endlessly.

**Comment by majocha** [comment]
I made the accessibility public for the whole thing for now, dealing with inline internal stuff across test projects is pita.

One usability advantage over the normal `async` I notice: there are no weird edge cases wrt `AsyncLocal` diagnostics loggers now. That's because the execution is driven by `AsyncTaskMethodBuilder`, we can be sure everything within the braces of `async2 { ... }` is isolated wrt to execution context.

**Comment by majocha** [comment]
Implementing fast tail call with `ReturnFromFinal` is tricky While we can abandon intermediate computations in a tail call chain, we cannot abandon the first one, we must return to it, because it is the one actually awaited.

One possible solution is to thread a `IValueTaskSource` context between tail calls. We can bind to a `ValueTask` in the initial `ReturnFromFinal`, set the value in the last call of the chain, dropping all the intermediate computations.

**Comment by majocha** [comment]
I went simply with TaskCompletionSource as a tail-call context. With `ReturnFromFinal` the overhead is now significantly lower.

**Comment by T-Gro** [comment]
(closing older PRs in bulk, feel free to reopen when reviving)
